sequencer/src/auction.rs

```diff
@@ -0,0 +1,103 @@
+use crate::{state::FeeAmount, NamespaceId, Transaction};
+use hotshot_types::data::ViewNumber;
+use std::num::NonZeroU64;
+
+/// State Machine for Auction
+#[derive(Clone, Copy)]
+struct Auction {
+    /// current phase.
+    phase: AuctionPhase,
+    /// A structure to find which Phase should be the current one.
+    // This could probably be an enum with some restructuring.
+    possible_phases: [AuctionPhase; 3],
+}
+
+impl Auction {
+    fn update(mut self, view: ViewNumber) {
+        self.phase = self
+            .possible_phases
+            .into_iter()
+            .find(|phase| (view > phase.start && view < phase.end))
+            .unwrap();
+    }
+    fn recv_bid(&self, view: ViewNumber, bid: BidTx) {
+        self.update(view);
+        match self.phase.kind {
+            AuctionPhaseKind::Bid => self.send_txn(bid.as_txn()),
+            _ => unimplemented!(),
+        }
+    }
+    fn send_txn(&self, txn: Transaction) {
+        unimplemented!();
+    }
+}
+
+// - needs to be configured in genesis block
+// - needs to be updatable
+/// Configuration for the auction system
+struct AuctionConfig {
```

Comment:
I can see in a very generic way why we need this information, but I'm having trouble reasoning about how we will actually use it. I feel like we will need to receive an event announcing the start of a phase. And from there we can use this to calculate the view at which that phase will end. Are there indeterminate spans between phases? if end of one phase == start of next then we would only need the start of the first phase on the chain, which could just be a constant (or come from config).